go/tools/releaser/file.go

```diff
@@ -0,0 +1,286 @@
+// Copyright 2021 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"compress/gzip"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+)
+
+var repoRootState = struct {
+	once sync.Once
+	dir  string
+	err  error
+}{}
+
+// repoRoot returns the workspace root directory. If this program was invoked
+// with 'bazel run', repoRoot returns the BUILD_WORKSPACE_DIRECTORY environment
+// variable. Otherwise, repoRoot walks up the directory tree and finds a
+// WORKSPACE file.
+func repoRoot() (string, error) {
+	repoRootState.once.Do(func() {
+		if wsDir := os.Getenv("BUILD_WORKSPACE_DIRECTORY"); wsDir != "" {
+			repoRootState.dir = wsDir
+			return
+		}
+		dir, err := os.Getwd()
+		if err != nil {
+			repoRootState.err = err
+			return
+		}
+		for {
+			_, err := os.Stat(filepath.Join(dir, "WORKSPACE"))
+			if err == nil {
+				repoRootState.dir = dir
+				return
+			}
+			if err != os.ErrNotExist {
+				repoRootState.err = err
+				return
+			}
+			parent := filepath.Dir(dir)
+			if parent == dir {
+				repoRootState.err = errors.New("could not find workspace directory")
+				return
+			}
+			dir = parent
+		}
+	})
+	return repoRootState.dir, repoRootState.err
+}
+
+// extractArchive extracts a zip or tar.gz archive opened in f, into the
+// directory dir, stripping stripPrefix from each entry before extraction.
+// name is the name of the archive, used for error reporting.
+func extractArchive(f *os.File, name, dir, stripPrefix string) (err error) {
```

Comment:
I would have expected this to return an [fs.FS](https://pkg.go.dev/io/fs#FS)

Reply:
Looking at `archive/tar` et al, looks like that might be a tall order. You could use use [`os.DirFS`](https://pkg.go.dev/os#DirFS) to provide this without much extra work.

Reply:
For this purpose, I really do need to extract the archive though, so that diff, patch, and gazelle can run in the extracted directories.